src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/SelfHostDeployer.cs

```diff
@@ -16,136 +18,158 @@ namespace Microsoft.AspNetCore.Server.IntegrationTesting
+                    WorkingDirectory = workingDirectory
+                };
+
+                AddEnvironmentVariablesToProcess(startInfo, DeploymentParameters.EnvironmentVariables);
+
+                var started = new TaskCompletionSource<string>();
+
+                HostProcess = new Process() { StartInfo = startInfo };
+                HostProcess.EnableRaisingEvents = true;
+                HostProcess.OutputDataReceived += (sender, dataArgs) =>
+                {
+                    if (!string.IsNullOrEmpty(dataArgs.Data))
+                    {
+                        var m = NowListeningRegex.Match(dataArgs.Data);
+                        if (m.Success)
+                        {
+                            started.TrySetResult(m.Groups["url"].Value);
+                        }
+                    }
+                };
+                var hostExitTokenSource = new CancellationTokenSource();
+                HostProcess.Exited += (sender, e) =>
+                {
+                    Logger.LogInformation("host process ID {pid} shut down", HostProcess.Id);
+                    TriggerHostShutdown(hostExitTokenSource);
+                };
 
-            var startInfo = new ProcessStartInfo
-            {
-                FileName = executableName,
-                Arguments = executableArgs,
-                UseShellExecute = false,
-                CreateNoWindow = true,
-                RedirectStandardError = true,
-                RedirectStandardOutput = true,
-                // Trying a work around for https://github.com/aspnet/Hosting/issues/140.
-                RedirectStandardInput = true,
-                WorkingDirectory = workingDirectory
-            };
-
-            AddEnvironmentVariablesToProcess(startInfo, DeploymentParameters.EnvironmentVariables);
-
-            HostProcess = new Process() { StartInfo = startInfo };
-            HostProcess.ErrorDataReceived += (sender, dataArgs) => { Logger.LogError(dataArgs.Data ?? string.Empty); };
-            HostProcess.OutputDataReceived += (sender, dataArgs) => { Logger.LogInformation(dataArgs.Data ?? string.Empty); };
-            HostProcess.EnableRaisingEvents = true;
-            var hostExitTokenSource = new CancellationTokenSource();
-            HostProcess.Exited += (sender, e) =>
-            {
-                TriggerHostShutdown(hostExitTokenSource);
-            };
+                try
+                {
+                    HostProcess.StartAndCaptureOutAndErrToLogger(executableName, Logger);
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError("Error occurred while starting the process. Exception: {exception}", ex.ToString());
+                }
 
-            try
-            {
-                HostProcess.Start();
-                HostProcess.BeginErrorReadLine();
-                HostProcess.BeginOutputReadLine();
-            }
-            catch (Exception ex)
```

Comment:
I recall adding this when the I couldn't get enough information from the logs as the process was crashing immediately. Could we add this in the new helper extension that you have in this PR?